Add ClickOnce publishing target for Dotnet app publishing

sujitnayak opened the pull request:
1. ClickOnce publishing needs to inject some of its targets into the publishing process. To do this, a extension point has been added to MS.Net,Publish.targets to import targets for any providers that need to plug into the publishing process. The extension point is based on the PublishProtocol property in the .pubxml file.
2. Adding a ClickOnce targets file that will get imported into the MS.Net.Publish.targets file and extend the publishing process by setting some custom properties and injecting its targets into the publishing target dependencies.

dsplaisted commented:
@sujitnayak Do you want this in .NET 5.0 / VS 16.8?  If so, the right branch to target right now is release/5.0.1xx.

sujitnayak committed:
Add extension point to MS.NET.Publish targets so that ClickOnce publish can inject it's targets during publishing process
The test included is disabled for now until ClickOnce changes in msbuild repo propagete to dotnet sdk.

sujitnayak commented:
> @sujitnayak Do you want this in .NET 5.0 / VS 16.8? If so, the right branch to target right now is release/5.0.1xx.

Correct, I have re-targetted the PR to release/5.0.1xx branch.

MSLukeWest commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ClickOnce.targets:
```diff
@@ -0,0 +1,77 @@
+  </ItemGroup>
+  
+  <ItemGroup Condition="'$(GenerateDependencyFile)' == 'true'">
+    <ProjectDepsFilesForClickOnce Include="$(ProjectDepsFilePath)"/>
```
Using ProjectDepsFilePath will be correct in the default new project scenario, but there are scenarios where a second publish specific deps file is created and you'll need to use that.  To reproduce this issue, create a new WPF .NET Core app, add a Nuget reference to "Microsoft.EntityFrameworkCore.Tools" and then build and run your app.  The app will likely fail to run, and in event viewer you'll see the reason is that there was a dependency that isn't found.

Fortunately there's already a property that tells you which deps file to use, and I just fixed this exact issue for wapproj so you can follow this example: https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/270214

Basically this will need to be changed to:

<ProjectDepsFilesForClickOnce Condition="'$(_UseBuildDependencyFile)' == 'true'" Include="$(ProjectDepsFilePath)"/>
<ProjectDepsFilesForClickOnce Condition="'$(_UseBuildDependencyFile)' != 'true'" Include="$(PublishDepsFilePath)"/>

And you may also need to depend on target _ComputeUseBuildDependencyFile

NikolaMilosavljevic replied:
It would be ideal if ClickOnce can reuse the fixes made for Installer Projects automatically, by consuming same output groups. That way a single fix would solve the publishing issue for all types. ClickOnce publishing is based on folder publishing, and I presume Installer Projects depend on the same.

sujitnayak replied:
Updated ClickOnce targets to use correct deps json file based on Luke's comment.

sujitnayak replied:
@NikolaMilosavljevic ClickOnce uses output groups that are visible in msbuild as well as Visual Studio for ClickOnce processing (e.g. Built, RefrenceCopyLocalPaths, Content etc). We have ClickOnce UI that also iterates through files in known output groups to show in the UI. Since we need to keep them in sync, it is risky to change the output groups we consume.

marcpopMSFT commented:
@sujitnayak Not that RC2 will close in roughly 2 weeks so this will need to get in by then.

wli3 commented:
Tests?

wli3 reviewed (dismissed):
No test

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -34,6 +34,8 @@ Copyright (c) .NET Foundation. All rights reserved.
     ============================================================
     -->
 
+  <Import Project="Microsoft.NET.$(PublishProtocol).targets" Condition="'$(PublishProtocol)' != '' and Exists('Microsoft.NET.$(PublishProtocol).targets')" />
```
This cannot give a good error message if there is a typo in PublishProtocol. A space in it could be worse

sujitnayak replied:
If there is a typo in the PublishProtocol, the ClickOnce targets file will not be found, the import will be skipped and ClickOnce publishing will not happen. That said, the PublishProtocol value is set by the Publish Provider in its schema file and not set by the user directly.

wli3 replied:
> If there is a typo in the PublishProtocol, the ClickOnce targets file will not be found, the import will be skipped and ClickOnce publishing will not happen. 

This experience is hard to let the user trace back to the typo.

But if it is not set by the user that is fine. Do we have many PublishProtocol? and they are mutual exclusive?

sujitnayak replied:
Yes there can be several publishing providers with their own unique protocol names and they are are mutually exclusive. The provider author specifies the protocol in a schema file (schema file is not directly available the to end user).

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ClickOnce.targets:
```diff
@@ -0,0 +1,86 @@
+    </ClickOncePublishDependsOn>
+  </PropertyGroup>
+
+  <Target Name="ClickOncePublish" Condition="'$(PublishableProject)'=='true'" DependsOnTargets="$(ClickOncePublishDependsOn)" />
```
is it done? What is the target to do the package generation?

sujitnayak replied:
ClickOncePublish is a pseudo target that has dependent targets set by ClickOncePublishDependsOn. These dependent targets are defined in the MS.Common.CurrentVersion.targets file. The PublishProtocolProviderTargets property is an extension point defined in MS.Common.CurrentVersion.targets and we set this property's value to ClickOncePublish. This way we inject the dependent targets that we need to invoke during the .NET Core Publish process.

sujitnayak commented:
> No test

The real publish targets are defined in MS.Common.CurrentVersion.targets file in the MSBuild repo. The ClickOnce targets file being added here will set up properties that changes some behaviors of the targets in MS.Common.CurrentVersion.targets. We will add tests in the MSBuild repo.

wli3 commented:
@sujitnayak could you add tests for the expected properties/itemgroups from SDK layer?

MSBuild does not update its sdk frequently. It will take a while to realize there is a bug from SDK(SDK->installer->msbuild run tests). And in general it is good to shorten the time to discover bugs.

sujitnayak commented:
> @sujitnayak could you add tests for the expected properties/itemgroups from SDK layer?
> 
> MSBuild does not update its sdk frequently. It will take a while to realize there is a bug from SDK(SDK->installer->msbuild run tests). And in general it is good to shorten the time to discover bugs.

The changes to the actual tasks and targets are in the msbuild repo and any test I add right now is not going to pass. Can I open a work item to track adding tests when both sets of changes are checked in and included in a VS build?

John-Hart commented:
@wli3 or anyone else have any other concerns? We're trying to get this in for out feature going into 16.8

wli3 commented:
I'll unblock this for now since we are in a rush. I don't know why we need msbuild side change. We just need to asset necessary properties and itemgroup that you just wrote. Please add tests before GA. If you merge this now. SDK team won't fix issues caused Microsoft.NET.ClickOnce.targets until tests are added

sujitnayak commented:
> I'll unblock this for now since we are in a rush. I don't know why we need msbuild side change. We just need to asset necessary properties and itemgroup that you just wrote. Please add tests before GA. If you merge this now. SDK team won't look into issues caused by Microsoft.NET.ClickOnce.targets until tests are added

@wli3 I have the test authored but the issue I am facing is that our publishing tasks need to run in the full framework instead of the .net core. For this release, we only support our publishing via VS and msbuild command line invocation. We will add support for the 'dotnet publish' scenario in the next release. Unless there is a way to invoke the clickonce test using straight msbuild.exe command line, I can only add the test when we add support for 'dotnet publish'.

wli3 commented:
Could you edit https://github.com/dotnet/sdk/blob/master/CODEOWNERS to include Microsoft.NET.ClickOnce.targets under your team github handle? As long as you can maintain it. It is also fine.

At the same time, use FullMSBuildOnly like https://github.com/dotnet/sdk/blob/2e1238421641e145b3db8d95985df948c79a1b95/src/Tests/Microsoft.NET.Build.Tests/COMReferenceTests.cs#L72 could test only in full msbuild

dsplaisted commented:
@sujitnayak You can mark your tests with `FullMSBuildOnlyFact` or `FullMSBuildOnlyTheory` to restrict them to running on full MSBuild only.  To test locally in VS, what I do is set TestCommandLine.UseFullFrameworkMSBuild to true.

sujitnayak commented:
> @sujitnayak You can mark your tests with `FullMSBuildOnlyFact` or `FullMSBuildOnlyTheory` to restrict them to running on full MSBuild only. To test locally in VS, what I do is set TestCommandLine.UseFullFrameworkMSBuild to true.

When I add FullMSBuildOnlyTheory, the test gets skipped when I run 'build -test' and I see this in the logs:

<test time="0" name="Microsoft.NET.Publish.Tests.GivenThatWeWantToPublishAClickOnceProject.It_publishes_with_a_publish_profile" result="Skip" method="It_publishes_with_a_publish_profile" type="Microsoft.NET.Publish.Tests.GivenThatWeWantToPublishAClickOnceProject">
<reason>
<![CDATA[This test requires Full MSBuild to run]]>
</reason>
</test>

Since I need to get this checked in, I am going to go ahead with the checkin and I will work on the test side with any follow up comments you may have.

dsplaisted commented:
@sujitnayak You can set the `TestFullMSBuild` environment variable to true to do a full framework test run.  Or you can modify the TestCommandLine class temporarily

sujitnayak commented:
> @sujitnayak You can set the `TestFullMSBuild` environment variable to true to do a full framework test run. Or you can modify the TestCommandLine class temporarily like I suggested

That worked. The test still fails because the msbuild bits under .tools\msbuild\16.8.0-preview3\tools\MSBuild\Current\Bin do not have my changes checked into the msbuild repo. I will keep the test disabled until msbuild bits get refreshed in dotnet/sdk.

dsplaisted commented:
@sujitnayak Did you mean to merge this instead of closing it?

sujitnayak commented:
> @sujitnayak Did you mean to merge this instead of closing it?

My mistake. I probably clicked the wrong button.
